### ISSUE TITLE
config-profiles not working?

### ISSUE THREAD

**Issue by omeganon** [synfinatic/aws-sso-cli#471]
**Output of `aws-sso version`:**
```
AWS SSO CLI Version 1.9.10 -- Copyright 2021-2022 Aaron Turner
11ec65288f78702a5e80078c04e0ef8b4ac7cb67 (v1.9.10) built at 2023-03-04T03:16:22+0000
```

**Describe the bug:**
After initial configuration and testing, I'm now testing the ability to refresh the account/roles list by assigning myself a new Role in an account in IAM Identity Center. I am able to see and select the new account/role combination from <mysso>.awsapps.com but I am unable to get aws-so command line to see the new account.

**To Reproduce:**
 1. View current assignments via 'aws-sso list'
 2. Assign new user to Account/Role in Identity Center...
 3. Refresh list with 'aws-sso config-profiles' (optionally with --force)
 4. View current assignments and see that the new role is not available via 'aws-sso list'

**Expected behavior:**
I expected the new Account/Role combination to be listed and usable. 

Thanks in advance!

**Screenshots:**
n/a

**Desktop (please complete the following information):**
 - OS: MacOS
 - Version 12.6.1

**Additional context:**
n/a

**Contents of `~/.aws-sso/config.yaml`:**
```
SSOConfig:
    myName<redacted>:
        SSORegion: us-east-1
        StartUrl: https://<redacted>.awsapps.com/start
        DefaultRegion: us-east-1
DefaultSSO: <redacted>
DefaultRegion: us-east-1
ConsoleDuration: 60
CacheRefresh: 1
Threads: 5
AutoConfigCheck: true
UrlAction: open
ConfigProfilesUrlAction: open
LogLevel: error
HistoryLimit: 10
HistoryMinutes: 1440
ProfileFormat: "{{ AccountIdStr .AccountId }}:{{ .RoleName }}"
AccountPrimaryTag:
    - AccountName
    - AccountAlias
    - Email
PromptColors:
    descriptionbgcolor: Turquoise
    descriptiontextcolor: Black
    inputbgcolor: DefaultColor
    inputtextcolor: DefaultColor
    prefixbackgroundcolor: DefaultColor
    prefixtextcolor: Blue
    previewsuggestionbgcolor: DefaultColor
    previewsuggestiontextcolor: Green
    scrollbarbgcolor: Cyan
    scrollbarthumbcolor: LightGrey
    selecteddescriptionbgcolor: DarkGray
    selecteddescriptiontextcolor: White
    selectedsuggestionbgcolor: DarkGray
    selectedsuggestiontextcolor: White
    suggestionbgcolor: Cyan
    suggestiontextcolor: White
ListFields:
    - AccountId
    - AccountAlias
    - RoleName
    - Profile
    - ExpiresStr```

**Comment by synfinatic** [comment]
Sounds like you're looking for `aws-sso cache`.  

TL;DR: `aws-sso` caches the list of roles because for large numbers of roles this can be very time consuming.  In your config you set `CacheRefresh: 1` so you've set it to 1hr.  https://github.com/synfinatic/aws-sso-cli/blob/main/docs/config.md#cacherefresh

**Comment by omeganon** [comment]
Thanks. I'm pretty sure that I did try aws-sso cache to clear the cache with no noticeable change. I'll try again. The use case I'm working against is that we have an external tool that allows for dynamic promotion to elevated roles within AWS and the expectation is that those are available immediately after that promotion. I'm trying to determine the steps necessary to get aws-sso to see those as soon as possible. Maybe it's `aws-sso cache && aws-sso config-profiles`...

**Comment by synfinatic** [comment]
You can disable the cache per the docs.  If `aws-sso` is not discovering the new role and while it is showing up in the AWS SSO role picker in the browser, then I'll have to dig in more, but so far from your description this sounds very much like the cache.  

Sadly, AWS doesn't provide a single API call to invalidate the entire cache or specific accounts. You have to first discover all the accounts and then get the list of roles for each account.  For orgs with many accounts this was actually taking a few minutes before I added parallelism, but we're still constrained by the AWS API rate limit.

**Comment by omeganon** [comment]
Sorry for the delay in getting back to this. I've been working on converting our accounts over to Identity Center and now have additional users who are experiencing this. I've tried the following and still have missing account/roles that are visible in the Identity center selector that are not present in aws-sso-cli (see screenshot). I am expecting to see a `-FullAdmin` role option.

![Screen Shot 2023-03-31 at 9 43 57 AM](https://user-images.githubusercontent.com/1200360/229154294-ae08ca46-5815-4fb1-8d79-84c4b746f7cc.png)

`$ aws-sso list | grep "[redacted] Production"
[redacted] | [redacted] Production                               | [redacted]-ReadOnly      | [redacted]-ReadOnly      |

$ aws-sso cache

$ aws-sso list | grep "[redacted] Production"
[redacted] | [redacted] Production                               | [redacted]-ReadOnly      | 3[redacted]-ReadOnly      |

$ aws-sso flush
INFO    Deleted cached AWS STS credentials for [redacted]

$ aws-sso list | grep "[redacted] Production"
[redacted] Production                               | [redacted]-ReadOnly      | [redacted]-ReadOnly      |

$ cat config.yaml

SSOConfig:
    [redacted]:
        SSORegion: us-east-1
        StartUrl: https://[redacted].awsapps.com/start
        DefaultRegion: us-east-1
DefaultRegion: us-east-1
ConsoleDuration: 60
CacheRefresh: 0
AutoConfigCheck: false
UrlAction: open
ConfigProfilesUrlAction: open
LogLevel: info
HistoryLimit: 10
HistoryMinutes: 1440
#ProfileFormat: "{{ .AccountAlias }}:{{ AccountIdStr .AccountId }}:{{ .RoleName }}"`

Additionally, I've rm'd ~/.aws-sso/cache.json and all of the above with no impact. Any further suggestions?

**Comment by synfinatic** [comment]
Are you comfortable privately sharing some debug logs with me?  Can you email them to me at _\<my GitHub handle\>_ @gmail.com? How many total accounts & roles have the impacted users been assigned?

**Comment by omeganon** [comment]
Yes, totally fine with that. What commands would you like me to run? For me, I'm associated with 39 role/account permutations. Most people are 1/3 that or less.

**Comment by synfinatic** [comment]
probably just: `aws-sso -L trace cache` followed by `aws-sso -L trace list` will help explain why deleting the cache doesn't fix the issue.  This sounds like a bug.

That said, I should also point out that your `config.yaml` explicitly says "do not autoupdate the cache" and so it will by design not automatically discover new roles that are assigned via AWS SSO:

```
CacheRefresh: 0
AutoConfigCheck: false
```

docs: https://github.com/synfinatic/aws-sso-cli/blob/main/docs/config.md#cacherefresh

Hence the original complaint you made is "as designed".  You'll need to select more appropriate values for your needs.

**Comment by omeganon** [comment]
Yeah, I had CacheRefresh: 1, AutoConfigCheck: true, but when reading the docs, setting CacheRefresh <0 implicitly disabled AutoConfigCheck, I made it explicit in the config. I'll send new traces for both.

**Comment by synfinatic** [comment]
Update: Problem seems to go away when user specifies `aws-sso cache --threads 1`

**Comment by synfinatic** [comment]
So I can't reproduce that at all.  If I make changes in IAM Identity Center, it's immediately available via the API.   That said, I don't have like 26 accounts either or maybe something is going strange in your account or AWS region?   I've attached two macOS binaries (one for Intel, the other ARM) to the ticket which have a special debug feature which does the necessary raw API calls to fetch the list of accounts and the roles.  It completely avoids the aws-sso cache and the more complexity for performance.  In theory, you should be able to add a role/permission set to a user and then immediately query the roles for that user.  If you don't see the role, then perhaps open a ticket with AWS and ask them why you seem to be seeing a delay?

Just need to run: `aws-sso list-sso-roles --level debug`

[aws-sso-1.9.10-darwin-amd64.zip](https://github.com/synfinatic/aws-sso-cli/files/11128028/aws-sso-1.9.10-darwin-amd64.zip)
[aws-sso-1.9.10-darwin-arm64.zip](https://github.com/synfinatic/aws-sso-cli/files/11128030/aws-sso-1.9.10-darwin-arm64.zip)

If you're sensitive about running random binaries from strangers, you can build the binary yourself using the code in this branch: https://github.com/synfinatic/aws-sso-cli/tree/debug-roles-cache and just typing `make`.

**Comment by omeganon** [comment]
Just seeing this. I'll respond with the output directly. Definitely looking like an AWS caching issue and I now have an open support case.

**Comment by omeganon** [comment]
well, the response from support isn't what I'd hoped. The 1 hour cache does exist and is currently not modifiable. You have to log out of the session to reset it. Is there a 'logout' equivalent command for aws-sso? 'flush' doesn't seem to be the right thing that they would need.

> Thanks for your patience on this case. 
>  
> During an active AWS IAM Identity Center session, a group membership of a user is cached, so if new group membership is introduced or removed, it will not be reflected.
> 
> Unfortunately, it is not possible to force a refresh, and currently it is fixed at 1 hour (as observed on your comments)
> 
> The workaround for this known issue, is forcing a logout using the command "aws sso logout" and then re-login. 
> 
> There is a feature request to modify the refresh rate of this cache.
> I have already attached your case with the request to raise the voice. Your use case will be used as a proof of requirement for adding this feature.

**Comment by synfinatic** [comment]
did you try `aws-sso list-sso-roles --level debug` as described above??  that has no cache.

**Comment by omeganon** [comment]
Yeah, I expect what they're telling me is your equivalent command `aws-sso flush -t all && aws-sso cache`. I _did_ get it to refresh a permission set in less than an hour through some combination of that and either being logged in or out of the web chooser session. Trying to recreate the unintentional order of operations that worked. It's not as simple as the two aws-sso commands above.

**Comment by synfinatic** [comment]
`aws-sso flush` flushes cached authentication tokens which is like a `logout`, but doesn't impact the cache of roles that `aws-sso` has. 

Hence you should only need `aws-sso cache`.  Anything else is a bug.

**Comment by synfinatic** [comment]
Hey @omeganon, just checking back in to see where you are and if you need further help or if I should close this ticket.

**Comment by fajpunk** [comment]
I am also running into this issue, with 10 accounts in my AWS Org.  To get `aws-sso` to see a new role that gets assigned (one that shows up on the AWS SSO login page), I have to:

1. `rm ~/.aws-sso/cache.json`
2. delete the `aws-sso-cli` item from the login `login` keychain in the Keychain Access app
3. Clear all of the `awsapps.com` cookies from my browser

Then, running `aws-sso` picks up the new profile.

### Output of aws-sso version:

AWS SSO CLI Version 1.9.10 -- Copyright 2021-2022 Aaron Turner
11ec65288f78702a5e80078c04e0ef8b4ac7cb67 (v1.9.10) built at 2023-03-03T22:10:24+0000

### Desktop (please complete the following information):

OS: MacOS
Version 13.3 (22E252)
Additional context:
n/a

### Contents of ~/.aws-sso/config.yaml:
```
# AUTO START
SSOConfig:
  DefaultSSO:
    SSORegion: us-west-2
    StartUrl: <redacted>
    DefaultRegion: us-west-2
DefaultRegion: us-west-2
ConsoleDuration: 720
CacheRefresh: 168
AutoConfigCheck: true
UrlAction: open
ConfigProfilesUrlAction: open
LogLevel: error
HistoryLimit: 10
HistoryMinutes: 1440
ProfileFormat: "{{ kebabcase .AccountAlias }}-{{ lower .RoleName }}"

# You can customize the behaviour of aws-sso:
# https://github.com/synfinatic/aws-sso-cli/blob/main/docs/config.md
# Just be sure to put your customizations after "AUTO END"

# AUTO END

UrlAction: open-url-in-container
UrlExecCommand:
  - /Applications/Firefox.app/Contents/MacOS/firefox
  - "%s"
```

The START and END delimiters are from a tool that we wrote to manage this config file for folks.

**Comment by synfinatic** [comment]
@fajpunk: This ticket is closed.  If you want help, then please open another one. 

Also notice you have `CacheRefresh: 168`.  So you've told `aws-sso` to **NOT discover new roles for up to 168 hours**... which might explain what you've got going on.  But yeah, we can discuss in another ticket.
